### ISSUE TITLE
Bug in `@program`

### ISSUE THREAD

**Issue by jpfairbanks** [epatters/Catlab.jl#102]
In the following script the last line raises an exception on the latest master. I'm trying to encode constant values of `R` as nullary functions and that seems to be the problem.

```julia
using Pkg
using Test

using Catlab
using Catlab.Doctrines
using Catlab.Syntax

using Catlab.Graphics
using Catlab.Graphics.ComposeWiringDiagrams

using Catlab.WiringDiagrams


wd(x) = to_wiring_diagram(x)

#display(x) = to_composejl(add_junctions!(wd(x)), direction=:horizontal, labels=true)
display(x::WiringDiagram) = to_graphviz(x, orientation=LeftToRight, labels=true)

using Catlab.Programs
#compile_expr(f, name=:acc_train, args=[:query], arg_types=[:Query])

import Catlab.WiringDiagrams: to_hom_expr

moduleof(p::Presentation) = typeof(first(generators(p))).name.module
to_hom_expr(p::Presentation, f::WiringDiagram) = to_hom_expr(moduleof(p), f)

@present Math(FreeAbelianBicategoryRelations) begin
    ℝ::Ob # Reals
    𝔹::Ob # Booleans
    
    one::Hom(munit(FreeAbelianBicategoryRelations.Ob), ℝ)
    two::Hom(munit(FreeAbelianBicategoryRelations.Ob), ℝ)
    mul::Hom(otimes(ℝ,ℝ), ℝ)
    inv::Hom(ℝ,ℝ)
    neg::Hom(ℝ,ℝ)
    pow::Hom(otimes(ℝ,ℝ),ℝ)
    log::Hom(otimes(ℝ,ℝ), ℝ)
    ln::Hom(ℝ,ℝ)
    leq::Hom(otimes(ℝ,ℝ), 𝔹)
    sq::Hom(ℝ, ℝ)
end

d = @program Math (x::ℝ, y::ℝ) begin
    x² = sq(x)
    y² = sq(y)
    z²  = mplus{ℝ}(x², y²)
    return leq(z², one())
end

to_hom_expr(Math, add_junctions(d))

d = @program Math (x::ℝ, y::ℝ) begin
    x² = pow(x, two())
    y² = pow(y, two())
    z²  = mplus{ℝ}(x², y²)
    return leq(z², one())
end

display(d)

to_hom_expr(Math, add_junctions(d))
```

The exception is 

```julia
AssertionError: all((wire.source.box == input_id(d) || wire.target.box == output_id(d) for wire = wires(d)))

Stacktrace:
 [1] to_hom_expr(::Type, ::Type, ::WiringDiagram{Catlab.Doctrines.AbelianBicategoryRelations.Hom}) at /Users/jfairbanks6/.julia/dev/Catlab/src/wiring_diagrams/Expressions.jl:84
 [2] to_hom_expr(::Module, ::WiringDiagram{Catlab.Doctrines.AbelianBicategoryRelations.Hom}) at /Users/jfairbanks6/.julia/dev/Catlab/src/wiring_diagrams/Expressions.jl:97
 [3] to_hom_expr(::Presentation{Symbol}, ::WiringDiagram{Catlab.Doctrines.AbelianBicategoryRelations.Hom}) at ./In[2]:22
 [4] top-level scope at In[27]:1
```

**Comment by epatters** [comment]
Ah yes, nullary operations, the bane of my existence. I dream of a Julia v2.0 that allows dispatch on return type.

**Comment by jpfairbanks** [comment]
Isn't the fact that we can't dispatch on return type a symptom of **Set** oriented thinking? If we were programming in **Rel** you wouldn't privilege domains over codomains like that. 

P.S. If only every application let you use the tex shortcuts for unicode (\bbR and such). It works great in text editors, jupyter and the REPL but the lack of github support is sad.

**Comment by epatters** [comment]
I never thought of it that way, but you're absolutely right. All the standard programming notation assumes you're dealing with functions, not relations, which is why using `@program` for a bicategory of relations is awkward.

**Comment by epatters** [comment]
Yet another tricky corner case in the reduction algorithm. Your code should work now after also replacing with `munit(FreeAbelianBicategoryRelations.Ob)` with `munit()`, in view of the fix to #106.
